### ISSUE TITLE
Workspace Symbols slow on large project

### ISSUE THREAD

**Issue by ashb** [microsoft/pyright#1997]
**Is your feature request related to a problem? Please describe.**

I work on a large project (Apache Airflow) and using the "workspace" symbols is so slow as to be unusable -- 72 seconds and 65s in two consecutive tests I just performed.

The issue is that this time happens _every_ time a symbol is searched for, making it essentially unusable in this workflow

**Describe the solution you'd like**
I'd like some kind of caching, so that I only have to pay this time cost (the first time per session) I assume it is parsing the entire workspace and and any files

**Additional context**

My (local, uncommitted) pyrightconfig.json file contains:

```
{
  "exclude": [
    "build/lib",
    "provider_packages/airflow",
    "provider_packages/build",
    ".eggs*",
    "build-*/**",
    "**/node_modules",
    "**/__pycache__",
  ],
  "reportUnusedImport": true,
  "reportDuplicateImport": true,
  "reportUnusedVariable": true,
  "reportUnknownParameterType": false,
  "reportUnboundVariable": true,
  "reportUnnecessaryIsInstance": true,
  "reportUnnecessaryCast": true,
  "venv": "airflow"
}
```

Codium (linux) extension (not vim/nvim this time) details:

Name: Pyright
Id: ms-pyright.pyright
Description: VS Code static type checking for Python
Version: 1.1.150
Publisher: ms-pyright

**Comment by erictraut** [comment]
Thanks for the feature request.

Pylance, the VS Code language server based on pyright, implements indexing functionality for this purpose. 
We don't have any plans to move that functionality to pyright, which is primarily focused on type checking.

**Comment by ashb** [comment]
@erictraut Cool, I understand that!

Happy for you to close this issue as a Wont Fix if this is not something you'd accept a PR for, or for it to stay open if you would accept a PR if I/someone writes it.

**Comment by smackesey** [comment]
Casting a vote for this feature: the unusably slow workspace symbol search is the biggest weakness of pyright. Also two questions:

- @ashb Did you find an alternative/workaround for this? All I want to do is quickly fuzzy-search to jump to classes/functions.
- @erictraut Doesn't completion do basically the same thing? For instance, there is a class `WorkspaceProcessContext` in my repo. When I type "Wo", pyright returns "WorkspaceProcessContext" in the completion candidate list almost immediately. But when I send a "workspace/symbol" request with the full name "WorkspaceProcessContext" as the query, pyright takes 35 seconds to return the result. How can symbol search in one context be so slow while it's so fast in the other?

**Comment by ashb** [comment]
No, I never found a work around (I even tried using pylance and vscode and it was unusably slow on apache/airflow repo)

**Comment by erictraut** [comment]
@smackasey, completion uses symbols that happen to have been seen already. If you have diagnosticMode set to "openFilesOnly", that means it will offer completions only for symbols that have been seen in files you've previously opened in the editor or the files they import. If you have diagnosticMode set to "workspace", it will include all files in your workspace, assuming that it's not an extremely large workspace. If it's too large, pyright will be forced to purge its cache to avoid exhausting the heap. So one suggestion is to try setting diagnosticMode to "workspace" and see if that improves your experience here.

The "Find all references" feature requires searching all of the files in the workspace. I recently added an optimization to this path that sped up the initial find by about 4x and subsequent finds by 50%. I accomplished this by simply loading the contents of the file and doing a quick text scan for the symbol and skipping tokenization, parsing and binding if the file didn't contain the target symbol. This optimization doesn't work for the "Workspace symbols" feature, however.

Pylance adds a background indexing feature that significantly improves the speed of some of these operations, but we don't have any plans to port this to pyright. As I've said before, pyright is focused on type checking. It does have basic language server features, but that's not its primary use. If you want these more advanced language server features, please consider switching to VS Code and Pylance.

**Comment by smackesey** [comment]
@erictraut Thank you for the detailed response, I've noticed that "Find all references" is considerably faster than `workspace/symbols` with the full name as query, so that makes sense-- thanks for that optimization.

From the first paragraph I am unclear on why setting diagnosticMode to "workspace" would be expected to help `workspace/symbols` performance-- from your explanation, it seems like this just controls the scope of completion candidates, which I have no issue with.

Unfortunately switching to VSCode/pylance is no option for me, I'm heavily invested in Neovim.

**Comment by erictraut** [comment]
Switching to diagnosticMode "workspace" causes pyright to load, analyze and cache the contents of all files in your project when you open it. As long as it's able to keep all of this information in memory, it can then perform a "Find all references" or "Workspace symbols" operation much more quickly based on this cached semantic information. 

If diagnosticMode is set to "openFilesOnly", pyright will perform the semantic analysis of all files in your workspace on demand, when you issue the first "Workspace symbols" command. Subsequent operations should be just as fast as if you had enabled "workspace" because the information will be cached in memory. So this suggestion will speed up only the first "Workspace symbols" operation. That's typically the one that user complain about because it requires a long and expensive operation.

**Comment by dam5h** [comment]
Given the `addressed in next version` tag, does this mean this will be implemented on pyright itself?  I am getting an empty response for results when I call `workspace/symbols`

**Comment by erictraut** [comment]
`workspace/symbols` is implemented in a limited form in pyright. For a full (and performant) implementation, an indexing facility is required. Pylance, which is Microsoft's Python language server that builds on Pyright, implements this functionality.

**Comment by dam5h** [comment]
Hi @erictraut !  Thanks for `pyright` it is such a great LSP for python.  I use helix as an editor so I don't have access to PyLance.

Oddly when I call `workplace/symbols` I just get an empty response immediately.  How would you suggest I go about debugging that from the `pyright` side?  I can also look into things on the helix side, but it does seem to be logging a response.  Is it possible that I have disabled `workspace/symbol` in my config somehow?

```
[tool.pyright]
venvPath = "/home/dave/.pyenv/versions"
venv="edgar-3.10"
pythonVersion = "3.10"
diagnosticMode="workspace"
```

```
2022-10-12T09:00:54.325 helix_term::ui::picker [DEBUG] picker score 221.289µs
2022-10-12T09:01:26.973 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"workspace/symbol","params":{"query":""},"id":2}
2022-10-12T09:01:26.974 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","id":2,"result":[]}
2022-10-12T09:01:26.974 helix_lsp::transport [INFO] <- []
2022-10-12T09:01:28.820 helix_term::ui::picker [DEBUG] picker score 810ns
``` 

Thank you for responding on this closed issue!

Cheers,
Dave

**Comment by erictraut** [comment]
Workspace symbols generally work if I use the pyright LSP in VS Code, so this is probably not an issue in pyright. If you can find a repro in VS Code, I can take a look at it. If you're using a different editor, you'll need to do some debugging on your own to figure out the problem. For instructions on building and debugging pyright, refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/build-debug.md).

**Comment by dam5h** [comment]
Okay got it, I will check it out, thanks!

**Comment by olejorgenb** [comment]
In case anyone else stumbles over this and looking for a clear answer to the `query: ""` case:

pyright explicitly disable listing all symbols:

```
    private _reportSymbolsForProgram(program: ProgramView) {
        // Don't do a search if the query is empty. We'll return
        // too many results in this case.
        if (!this._query) {
            return;
        }
```
